File: surface/widgets/appkit/components/widgets/concepts/governance-decision/ProposalCard.swift
import AppKit

class ProposalCardView: NSView {

    enum State: String { case idle; case hovered; case focused; case navigating }

    private(set) var state: State = .idle
    private var title: String = ""
    private var descriptionText: String = ""
    private var author: String = ""
    private var proposalStatus: String = ""
    private var timestamp: String = ""
    private var variant: String = "full"
    private var showVoteBar: Bool = true
    private var showQuorum: Bool = false
    private var truncateLength: Int = 120
    private var trackingArea: NSTrackingArea?

    var onClick: (() -> Void)?
    var onNavigate: (() -> Void)?

    private let rootStack = NSStackView()
    private let statusBadge = NSTextField(labelWithString: "")
    private let titleLabel = NSTextField(labelWithString: "")
    private let descriptionLabel = NSTextField(wrappingLabelWithString: "")
    private let proposerLabel = NSTextField(labelWithString: "")
    private let voteBarPlaceholder = NSView()
    private let quorumPlaceholder = NSView()
    private let timeRemainingLabel = NSTextField(labelWithString: "")
    private let actionButton = NSButton(title: "View", target: nil, action: nil)

    // MARK: - State machine

    func reduce(_ event: String) {
        switch state {
        case .idle:
            if event == "HOVER" { state = .hovered }
            if event == "FOCUS" { state = .focused }
            if event == "CLICK" { state = .navigating; performNavigation() }
        case .hovered:
            if event == "UNHOVER" { state = .idle }
        case .focused:
            if event == "BLUR" { state = .idle }
            if event == "CLICK" || event == "ENTER" { state = .navigating; performNavigation() }
        case .navigating:
            if event == "NAVIGATE_COMPLETE" { state = .idle }
        }
        updateUI()
    }

    // MARK: - Init

    override init(frame: NSRect) {
        super.init(frame: frame)
        setupSubviews()
        setAccessibilityRole(.group)
        setAccessibilityLabel("Proposal card")
    }

    required init?(coder: NSCoder) { super.init(coder: coder) }

    // MARK: - Setup

    private func setupSubviews() {
        wantsLayer = true
        layer?.cornerRadius = 8
        layer?.borderWidth = 1
        layer?.borderColor = NSColor.separatorColor.cgColor

        rootStack.orientation = .vertical
        rootStack.spacing = 6
        rootStack.translatesAutoresizingMaskIntoConstraints = false
        addSubview(rootStack)
        NSLayoutConstraint.activate([
            rootStack.topAnchor.constraint(equalTo: topAnchor, constant: 12),
            rootStack.leadingAnchor.constraint(equalTo: leadingAnchor, constant: 12),
            rootStack.trailingAnchor.constraint(equalTo: trailingAnchor, constant: -12),
            rootStack.bottomAnchor.constraint(equalTo: bottomAnchor, constant: -12),
        ])

        // Status badge
        statusBadge.font = .boldSystemFont(ofSize: 11)
        statusBadge.wantsLayer = true
        statusBadge.layer?.cornerRadius = 4
        statusBadge.setAccessibilityRole(.staticText)
        rootStack.addArrangedSubview(statusBadge)

        // Title
        titleLabel.font = .boldSystemFont(ofSize: 15)
        titleLabel.maximumNumberOfLines = 2
        titleLabel.setAccessibilityRole(.staticText)
        rootStack.addArrangedSubview(titleLabel)

        // Description
        descriptionLabel.font = .systemFont(ofSize: 13)
        descriptionLabel.textColor = .secondaryLabelColor
        descriptionLabel.maximumNumberOfLines = 3
        rootStack.addArrangedSubview(descriptionLabel)

        // Proposer
        proposerLabel.font = .systemFont(ofSize: 12)
        proposerLabel.textColor = .tertiaryLabelColor
        rootStack.addArrangedSubview(proposerLabel)

        // Vote bar placeholder
        voteBarPlaceholder.heightAnchor.constraint(equalToConstant: 24).isActive = true
        rootStack.addArrangedSubview(voteBarPlaceholder)

        // Quorum placeholder
        quorumPlaceholder.heightAnchor.constraint(equalToConstant: 20).isActive = true
        rootStack.addArrangedSubview(quorumPlaceholder)

        // Time remaining
        timeRemainingLabel.font = .monospacedDigitSystemFont(ofSize: 12, weight: .regular)
        timeRemainingLabel.textColor = .secondaryLabelColor
        rootStack.addArrangedSubview(timeRemainingLabel)

        // Action button
        actionButton.bezelStyle = .roundRect
        actionButton.target = self
        actionButton.action = #selector(handleActionClick(_:))
        actionButton.setAccessibilityLabel("View proposal")
        rootStack.addArrangedSubview(actionButton)

        updateUI()
    }

    // MARK: - Configure

    func configure(title: String, description: String, author: String, status: String, timestamp: String, variant: String = "full", showVoteBar: Bool = true, showQuorum: Bool = false, truncateLength: Int = 120) {
        self.title = title
        self.descriptionText = description
        self.author = author
        self.proposalStatus = status
        self.timestamp = timestamp
        self.variant = variant
        self.showVoteBar = showVoteBar
        self.showQuorum = showQuorum
        self.truncateLength = truncateLength
        updateUI()
    }

    private func truncatedDescription() -> String {
        if descriptionText.count <= truncateLength { return descriptionText }
        return String(descriptionText.prefix(truncateLength).trimmingCharacters(in: .whitespaces)) + "\u{2026}"
    }

    private func actionLabel() -> String {
        switch proposalStatus {
        case "Active": return "Vote"
        case "Passed", "Approved": return "Execute"
        case "Draft": return "Edit"
        default: return "View"
        }
    }

    private func formatTimeRemaining() -> String {
        guard let date = ISO8601DateFormatter().date(from: timestamp) else { return timestamp }
        let diff = date.timeIntervalSinceNow
        let absDiff = abs(diff)
        let suffix = diff < 0 ? " ago" : " remaining"
        let days = Int(absDiff / 86400)
        let hours = Int(absDiff / 3600)
        let minutes = Int(absDiff / 60)
        if days > 0 { return "\(days)d\(suffix)" }
        if hours > 0 { return "\(hours)h\(suffix)" }
        return "\(minutes)m\(suffix)"
    }

    private func performNavigation() {
        onClick?()
        onNavigate?()
        DispatchQueue.main.async { [weak self] in
            self?.reduce("NAVIGATE_COMPLETE")
        }
    }

    private func updateUI() {
        let isMinimal = variant == "minimal"
        let isFull = variant == "full"

        statusBadge.stringValue = proposalStatus
        titleLabel.stringValue = title
        descriptionLabel.stringValue = truncatedDescription()
        descriptionLabel.isHidden = isMinimal
        proposerLabel.stringValue = "by \(author)"
        proposerLabel.isHidden = isMinimal
        voteBarPlaceholder.isHidden = !(showVoteBar && proposalStatus == "Active" && !isMinimal)
        quorumPlaceholder.isHidden = !(showQuorum && isFull)
        timeRemainingLabel.stringValue = formatTimeRemaining()
        actionButton.title = actionLabel()
        actionButton.isHidden = isMinimal
        actionButton.setAccessibilityLabel("\(actionLabel()) proposal: \(title)")

        // Status badge color
        switch proposalStatus {
        case "Active": statusBadge.textColor = .systemBlue
        case "Passed", "Approved": statusBadge.textColor = .systemGreen
        case "Rejected": statusBadge.textColor = .systemRed
        case "Cancelled": statusBadge.textColor = .secondaryLabelColor
        default: statusBadge.textColor = .labelColor
        }

        // Hover state
        if state == .hovered {
            layer?.backgroundColor = NSColor.controlBackgroundColor.withAlphaComponent(0.5).cgColor
        } else {
            layer?.backgroundColor = NSColor.clear.cgColor
        }

        setAccessibilityValue("\(proposalStatus) proposal: \(title)")
    }

    // MARK: - Actions

    @objc private func handleActionClick(_ sender: NSButton) {
        reduce("CLICK")
    }

    // MARK: - Mouse tracking

    override func updateTrackingAreas() {
        super.updateTrackingAreas()
        if let area = trackingArea { removeTrackingArea(area) }
        trackingArea = NSTrackingArea(rect: bounds, options: [.mouseEnteredAndExited, .activeInKeyWindow], owner: self, userInfo: nil)
        addTrackingArea(trackingArea!)
    }

    override func mouseEntered(with event: NSEvent) { reduce("HOVER") }
    override func mouseExited(with event: NSEvent) { reduce("UNHOVER") }
    override func mouseDown(with event: NSEvent) { reduce("CLICK") }

    // MARK: - Keyboard

    override var acceptsFirstResponder: Bool { true }

    override func becomeFirstResponder() -> Bool {
        reduce("FOCUS")
        return super.becomeFirstResponder()
    }

    override func resignFirstResponder() -> Bool {
        reduce("BLUR")
        return super.resignFirstResponder()
    }

    override func keyDown(with event: NSEvent) {
        switch event.keyCode {
        case 36, 49: reduce("ENTER") // Enter, Space
        default: super.keyDown(with: event)
        }
    }

    // MARK: - Cleanup

    deinit {}
}

File: surface/widgets/vue/components/widgets/concepts/governance-decision/ProposalCard.ts
import { defineComponent, h, ref, computed, onMounted, onUnmounted, watch } from 'vue';

/* ---------------------------------------------------------------------------
 * ProposalCard state machine
 * States: idle (initial), hovered, focused, navigating
 * ------------------------------------------------------------------------- */

export type ProposalCardState = 'idle' | 'hovered' | 'focused' | 'navigating';
export type ProposalCardEvent =
  | { type: 'HOVER' }
  | { type: 'UNHOVER' }
  | { type: 'FOCUS' }
  | { type: 'BLUR' }
  | { type: 'CLICK' }
  | { type: 'ENTER' }
  | { type: 'NAVIGATE_COMPLETE' };

export function proposalCardReducer(state: ProposalCardState, event: ProposalCardEvent): ProposalCardState {
  switch (state) {
    case 'idle':
      if (event.type === 'HOVER') return 'hovered';
      if (event.type === 'FOCUS') return 'focused';
      if (event.type === 'CLICK') return 'navigating';
      return state;
    case 'hovered':
      if (event.type === 'UNHOVER') return 'idle';
      return state;
    case 'focused':
      if (event.type === 'BLUR') return 'idle';
      if (event.type === 'CLICK') return 'navigating';
      if (event.type === 'ENTER') return 'navigating';
      return state;
    case 'navigating':
      if (event.type === 'NAVIGATE_COMPLETE') return 'idle';
      return state;
    default:
      return state;
  }
}

/* ---------------------------------------------------------------------------
 * Helpers
 * ------------------------------------------------------------------------- */

function truncate(text: string, max: number): string {
  if (text.length <= max) return text;
  return text.slice(0, max).trimEnd() + '\u2026';
}

function formatTimeRemaining(timestamp: string): string {
  const target = new Date(timestamp).getTime();
  const now = Date.now();
  const diffMs = target - now;
  const absDiff = Math.abs(diffMs);
  const seconds = Math.floor(absDiff / 1000);
  const minutes = Math.floor(seconds / 60);
  const hours = Math.floor(minutes / 60);
  const days = Math.floor(hours / 24);
  const isPast = diffMs < 0;
  const suffix = isPast ? ' ago' : ' remaining';
  if (days > 0) return `${days}d${suffix}`;
  if (hours > 0) return `${hours}h${suffix}`;
  if (minutes > 0) return `${minutes}m${suffix}`;
  return `${seconds}s${suffix}`;
}

function actionLabelForStatus(status: string): string {
  switch (status) {
    case 'Active': return 'Vote';
    case 'Passed':
    case 'Approved': return 'Execute';
    case 'Draft': return 'Edit';
    default: return 'View';
  }
}

/* ---------------------------------------------------------------------------
 * Component
 * ------------------------------------------------------------------------- */

export const ProposalCard = defineComponent({
  name: 'ProposalCard',
  props: {
    title: { type: String, required: true },
    description: { type: String, required: true },
    author: { type: String, required: true },
    status: { type: String, required: true },
    timestamp: { type: String, required: true },
    variant: { type: String, default: 'full' },
    showVoteBar: { type: Boolean, default: true },
    showQuorum: { type: Boolean, default: false },
    truncateDescription: { type: Number, default: 120 },
  },
  emits: ['click', 'navigate'],
  setup(props, { slots, emit }) {
    const state = ref<ProposalCardState>('idle');
    const send = (event: ProposalCardEvent) => {
      state.value = proposalCardReducer(state.value, event);
    };

    const truncatedDescription = computed(() => truncate(props.description, props.truncateDescription));
    const relativeTime = computed(() => formatTimeRemaining(props.timestamp));
    const actionLabel = computed(() => actionLabelForStatus(props.status));

    const showDescription = computed(() => props.variant !== 'minimal');
    const showProposer = computed(() => props.variant !== 'minimal');
    const showVoteBarSlot = computed(() => props.showVoteBar && props.status === 'Active' && props.variant !== 'minimal');
    const showQuorumSlot = computed(() => props.showQuorum && props.variant === 'full');
    const showAction = computed(() => props.variant !== 'minimal');

    // Navigate side-effect
    let navigateTimer: ReturnType<typeof setTimeout> | null = null;
    watch(state, (newState) => {
      if (newState === 'navigating') {
        emit('click');
        emit('navigate');
        navigateTimer = setTimeout(() => send({ type: 'NAVIGATE_COMPLETE' }), 0);
      }
    });
    onUnmounted(() => {
      if (navigateTimer) clearTimeout(navigateTimer);
    });

    const handleKeyDown = (e: KeyboardEvent) => {
      if (e.key === 'Enter' || e.key === ' ') {
        e.preventDefault();
        send({ type: state.value === 'focused' ? 'ENTER' : 'CLICK' });
      }
    };

    return () => h('article', {
      role: 'article',
      'aria-label': `${props.status} proposal: ${props.title}`,
      'data-surface-widget': '',
      'data-widget-name': 'proposal-card',
      'data-part': 'root',
      'data-state': state.value,
      'data-variant': props.variant,
      'data-status': props.status,
      tabindex: 0,
      onClick: () => send({ type: 'CLICK' }),
      onMouseenter: () => send({ type: 'HOVER' }),
      onMouseleave: () => send({ type: 'UNHOVER' }),
      onFocus: () => send({ type: 'FOCUS' }),
      onBlur: () => send({ type: 'BLUR' }),
      onKeydown: handleKeyDown,
    }, [
      // Status badge
      h('div', {
        'data-part': 'status-badge',
        'data-status': props.status,
        role: 'status',
        'aria-label': `Status: ${props.status}`,
      }, props.status),

      // Title
      h('h3', { 'data-part': 'title', role: 'heading', 'aria-level': 3 }, props.title),

      // Description
      showDescription.value
        ? h('p', {
            'data-part': 'description',
            'data-visible': props.variant !== 'minimal' ? 'true' : 'false',
          }, truncatedDescription.value)
        : null,

      // Proposer
      showProposer.value
        ? h('div', {
            'data-part': 'proposer',
            'data-author': props.author,
            'aria-label': `Proposed by ${props.author}`,
          }, [
            h('span', { 'data-part': 'avatar', 'aria-hidden': 'true' }),
            h('span', null, props.author),
          ])
        : null,

      // Vote bar slot
      showVoteBarSlot.value
        ? h('div', { 'data-part': 'vote-bar', 'data-visible': 'true' })
        : null,

      // Quorum gauge slot
      showQuorumSlot.value
        ? h('div', { 'data-part': 'quorum-gauge', 'data-visible': 'true' })
        : null,

      // Time remaining
      h('span', {
        'data-part': 'time-remaining',
        'data-timestamp': props.timestamp,
        role: 'timer',
        'aria-live': 'off',
      }, relativeTime.value),

      // Action button
      showAction.value
        ? h('button', {
            type: 'button',
            'data-part': 'action',
            role: 'button',
            'aria-label': `View proposal: ${props.title}`,
            tabindex: 0,
            onClick: (e: MouseEvent) => {
              e.stopPropagation();
              send({ type: 'CLICK' });
            },
          }, slots.default ? slots.default() : actionLabel.value)
        : null,
    ]);
  },
});

export default ProposalCard;
